sphinx-mode.el

```diff
@@ -36,26 +40,61 @@
   '((t (:inherit fixed-pitch)))
   "Face used for code blocks.")
 
-(defun sphinx-fontify-code-block (limit)
-  "Fontify code blocks from point to LIMIT."
-  (condition-case nil
-      (while (re-search-forward "\.\. code-block:: \\(.*?\\)\n\n\\( +\\)" limit t)
-        (let* ((block-start (match-end 0))
-               (block-highlight-start (match-beginning 2))
-               (lang (match-string 1))
-               (prefix (match-string 2))
-               (prefix-search (concat "^" prefix))
-               block-end)
-          (while (progn
-                   (forward-line)
-                   (looking-at prefix-search)))
-          (setq block-end (point))
-          (sphinx-src-font-lock-fontify-block lang block-start block-end)
-          (add-face-text-property
-           block-highlight-start block-end
-           'sphinx-code-block-face 'append)))
-    (error nil)))
+(defvar sphinx-code-block-default-language "python"
+  "The default language used for highlighting code blocks.")
 
+(defun sphinx--fontify-code-blocks (limit)
+  "Fontify code blocks from point to LIMIT."
+  (save-excursion
+    ;; This is for the ‘.. code-block::’ syntax.
+    (while (re-search-forward "\.\. code-block::" limit t)
+      (let* ((language
+              (s-trim (buffer-substring-no-properties (point) (line-end-position))))
+             (end (sphinx--fontify-block (point) language)))
+        (goto-char end))))
+  (save-excursion
+    ;; This is for the ‘Here is some code::’ syntax.
+    (while (re-search-forward "::$" limit t)
```

Comment:
Also probably this should be split into a separate function and we would then add both of them to the keywords variable in the mode switch